File: mobile/lib/features/products/models/barcode_lookup.dart
class BarcodeLookupProduct {
  final int? id;
  final String? barcode;
  final String? name;
  final String? brand;
  final String? spec;
  final String? manufacturer;
  final String? imageUrl;

  const BarcodeLookupProduct({
    this.id,
    this.barcode,
    this.name,
    this.brand,
    this.spec,
    this.manufacturer,
    this.imageUrl,
  });

  factory BarcodeLookupProduct.fromJson(Map<String, dynamic>? json) {
    if (json == null) return const BarcodeLookupProduct();
    return BarcodeLookupProduct(
      id: json['id'] as int?,
      barcode: json['barcode'] as String?,
      name: json['name'] as String?,
      brand: json['brand'] as String?,
      spec: json['spec'] as String?,
      manufacturer: json['manufacturer'] as String?,
      imageUrl: json['image_url'] as String?,
    );
  }
}

class BarcodeLookupResult {
  final bool found;
  final String? source;
  final BarcodeLookupProduct product;
  final List<String> errors;

  const BarcodeLookupResult({
    required this.found,
    this.source,
    this.product = const BarcodeLookupProduct(),
    this.errors = const [],
  });

  factory BarcodeLookupResult.fromJson(Map<String, dynamic> json) {
    return BarcodeLookupResult(
      found: json['found'] as bool? ?? false,
      source: json['source'] as String?,
      product: BarcodeLookupProduct.fromJson(
        json['product'] is Map
            ? Map<String, dynamic>.from(json['product'] as Map)
            : null,
      ),
      errors: (json['errors'] as List?)
              ?.map((error) => error.toString())
              .toList() ??
          const [],
    );
  }
}

File: frontend/src/utils/barcodeLookup.ts
import { api } from '@/api'
import { t } from '@/plugins/i18n'

export interface BarcodeLookupProduct {
  id?: number
  barcode?: string | null
  name?: string | null
  brand?: string | null
  spec?: string | null
  manufacturer?: string | null
  image_url?: string | null
}

export interface BarcodeLookupResult {
  found: boolean
  source: string | null
  product: BarcodeLookupProduct
  errors: string[]
  has_enabled_providers: boolean
}

export async function lookupBarcode(barcode: string): Promise<BarcodeLookupResult> {
  const code = barcode.trim()
  if (!code) {
    return { found: false, source: null, product: {}, errors: [t('barcodeLookup.errors.barcodeRequired')], has_enabled_providers: true }
  }

  try {
    const result = await api.get(`/products/entity/barcode/${encodeURIComponent(code)}`)
    if (result && typeof result.found === 'boolean') return result
  } catch (error: any) {
    if (error?.found === false) return error
    if (error?.response?.data?.found === false) return error.response.data
    return {
      found: false,
      source: null,
      product: {},
      errors: [error?.userMessage || t('barcodeLookup.errors.lookupFailed')],
      has_enabled_providers: false,
    }
  }
  return { found: false, source: null, product: {}, errors: [t('barcodeLookup.errors.invalidResponse')], has_enabled_providers: true }
}
